### PR TITLE
Defender 1.4 & Check for Sensitive Users, Groups, and Domains

### PR DESCRIPTION
**Pull request by Sloane4**
#  #

## 🗣 Description ##

Check Users, groups, & domains for Strict EOP against Config file.

## 💭 Motivation and context ##

- Confirms Strict EOP exists
- User can indicate Users, groups, & domains that must be protected
- Still passes if there are more users, groups, & domains protected than indicated in the config file.
- closing #486 

## 🧪 Testing ##

- Unit Tests
- Cisanet

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

- [ ] Create a release.

### CODE REVIEW

File: sample-config-files/defender-config.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  are present.  These parameters support operational use cases for having
+  backup or break glass account exclusions to global user policies without
+  failing best practices.  Any exemptions and their risks should be carefully
+  considered and documented as part of an organization's cybersecurity risk
+  management program process and practices.
+  This example includes the following design choices for illustrative purposes:
+  - Use of Pascal case convention for varible names
+  - Defines a namespace for values to be used across baselines/products (i.e., GlobalVars)
+  - Per product namespace for values related to that specific product (i.e., Aad, SharePoint)
+  - Namespace for each policy item within a product for variables related only to one policy item (i.e., Aad.Policy2_1)
```
[the remark this answers is not in the document]
Comment (Sloane4) [reply sample-config-files/defender-config.yaml]: @schrolla please confirm & resolve

File: Rego/DefenderConfig.rego [hunk 2]
```diff
@@ -61,6 +61,146 @@ ApplyLicenseWarning(Message) := concat("", [ReportDetailsBoolean(false), License
...
+##########################################
+
+SensitiveUsers(Policies, PolicyID) := {
+    "ConfigUsers" : ConfigUsers,
+    "ExcludedUsers" : ExcludedUsers,
+    "IncludedUsers" : IncludedUsers
+} {
+    Config := input.scuba_config.Defender[PolicyID].SensitiveIDs.Users
+    Policy := [Policy | Policy := Policies[_]; Policy.Identity == "Strict Preset Security Policy"]
+    ConfigUsers := { x | x := Config[_] }
```
[the remark this answers is not in the document]
Comment (Sloane4) [reply Rego/DefenderConfig.rego]: @schrolla please confirm & resolve

File: Rego/DefenderConfig.rego [hunk 3]
```diff
@@ -61,6 +61,146 @@ ApplyLicenseWarning(Message) := concat("", [ReportDetailsBoolean(false), License
...
+}
+
+SensitiveDomains(Policies, PolicyID) := {
+    "ConfigDomains" : ConfigDomains,
+    "ExcludedDomains" : ExcludedDomains,
+    "IncludedDomains" : IncludedDomains
+} {
+    Config := input.scuba_config.Defender[PolicyID].SensitiveIDs.Domains
+    Policy := [Policy | Policy := Policies[_]; Policy.Identity == "Strict Preset Security Policy"]
+    ConfigDomains := { x | x := Config[_] }
```
[the remark this answers is not in the document]
Comment (Sloane4) [reply Rego/DefenderConfig.rego]: @schrolla please confirm & resolve

File: Rego/DefenderConfig.rego [hunk 4]
```diff
@@ -61,6 +61,146 @@ ApplyLicenseWarning(Message) := concat("", [ReportDetailsBoolean(false), License
...
+    count([Policy | Policy = Policies[_];
+        Policy.Identity == "Strict Preset Security Policy";
+        Policy.SentToMemberOf == null;
+        Policy.ExceptIfSentToMemberOf == null]) > 0
+}
+
+GroupSensitiveIDs(Policies, PolicyID) := true if {
+    Config := input.scuba_config.Defender[PolicyID].SensitiveIDs.Groups
+    Policy := [Policy | Policy := Policies[_]; Policy.Identity == "Strict Preset Security Policy"]
+    ConfigGroups := { x | x := Config[_] }
```
[the remark this answers is not in the document]
Comment (Sloane4) [reply Rego/DefenderConfig.rego]: @schrolla please confirm & resolve

File: Testing/Unit/Rego/Defender/DefenderConfig_01_test.rego
[the remark this answers is not in the document]
Comment (Sloane4) [reply Testing/Unit/Rego/Defender/DefenderConfig_01_test.rego:1]: @schrolla please confirm & resolve, I added a few more too...

**Comment by adhilto** [comment]
@Sloane4 I just saw that you marked this as a draft. Should I hold off on reviewing?

**Comment by Sloane4** [comment]
@adhilto, yeah I made an incorrect assumption & the fix causes a big change so I have it in draft until the change is finished.

**Comment by Sloane4** [comment]
@adhilto & @schrolla ready for review

**Review by adhilto — APPROVED** [review]
Great work!

**Comment by Sloane4** [comment]
@nanda-katikaneni ready for merge

**Pull request merged**